src/workerd/api/sockets.c++

```diff
@@ -104,16 +104,20 @@ jsg::Ref<Socket> setupSocket(
     fulfiller->fulfill(true);
   });
 
-  auto watchForDisconnectTask = connection->whenWriteDisconnected()
-      .then([&disconnectedFulfiller]() {
-    disconnectedFulfiller.fulfill(false);
-  }, [&disconnectedFulfiller](kj::Exception&& e) {
-    disconnectedFulfiller.reject(kj::mv(e));
-  }).eagerlyEvaluate(
-        [deferredCancelDisconnected = kj::mv(deferredCancelDisconnected)](kj::Exception&& e) {
-    // Should never get here because we caught exceptions above.
-    LOG_EXCEPTION("socketWatchForDisconnectTask", e);
-  });
+  static auto constexpr handleDisconnected =
+      [](kj::Promise<void> disconnected, auto& fulfiller) -> kj::Promise<void> {
```

Comment:
I would use a Connection parameter instead of a disconnect promise and did (`co_await connection->whenWriteDisconnected`). Passing Promise as argument is a bit weird?